c++/src/capnp/schema.capnp

```diff
@@ -192,9 +192,16 @@ struct Node {
     struct Member {
       docComment @0 :Text;
       # Doc comment on the member.
+      bytePosition @1 :BytePosition;
```

Comment:
I'm not sure I understand how this would be used. It looks like you're filling this in with the position of the entire member declaration -- the whole line. So like if you have a field declaration:

```
struct Foo {
  bar @0 :Bar;
}
```

The `bytePosition` for `Foo.bar` would specify the entire text `bar @0 :Bar`?

But, when would you use this? I think the vast majority of the time, jump-to-definition wants to jump to a type name, not a member?

Or are you actually implementing jump-to-definition for things like:

`const foo :Foo = (bar = (i = 2));`

Here, I suppose someone might want to put the cursor on `bar` and jump-to-definition?

But this feels like a fairly unusual thing. More often I think people want to jump to type definitions.

That is, when we have the field declaration `bar @0 :Bar;`, you might put your cursor specifically on the type, `Bar`, and jump-to-definition from there. But how will the language server know that `Bar` is a type declaration here?

Perhaps instead of `bytePosition` as you've defined it, you really want something like `typeBytePosition`, which is the byte position of the _type_?

Of course, the type itself might be an expression, like `SomeModule.ParentType.Bar` or `SomeGeneric(Foo, Bar)`. So there might be multiple symbols to jump _from_. How should we represent this?

I kind of wonder if this is the right data structure for this. Maybe we really need to have the compiler output an additional list of all symbols that can be jumped-to, like:

```
struct Identifier {
  startByte @0 :UInt32;
  endByte @1 :UInt32;

  union {
    typeId @2 :UInt64;
    # Identifier refers to a type. This is the type ID.

    member :group {
      # Identifier refers to a member of a type.
      parentTypeId @3 :UInt64;
      ordinal @4 :UInt16;
    }
  }
}

identifiers :List(Identifier);
# Positions of all identifiers in the file, and what they refer to.
```

Reply:
@kentonv 

Thank you so much! I fully agree with your proposal.
I will replace the schema definition with your suggestion.  
Could you give me a hint on which files I should modify to fill in this information? In my understanding, modifying the node translator alone doesn't seem sufficient.

Reply:
So the trick is, everywhere where we are "resolving" an identifier, we need to make note of the identifier location and what it resolved to.

I think the vast majority of cases boil down to `BrandScope::compileDeclExpression()`. Here, we call `resolver.resolve()` or similar functions to handle RELATIVE_NAME, ABSOLUTE_NAME, and IMPORT expression types. In each case, we have a `LocatedText` that we're using as the input. You can see where we use `errorReporter.addErrorOn()` to report errors -- and the error of course reports exactly where the identifier was, so this proves we have the info we need.

The only question is: Where do we report the resolution result, so that it ends up in the identifier table in the source info?

I think the best way would be do add a method to the `ErrorReporter` interface, like:

```
virtual void reportResolution(uint32_t startByte, uint32_t endByte, uint64_t typeId) {}
```

(Later on we might need to expand the third parameter to support resolving to non-type members and to generic type parameters, but let's not worry about those for now (don't report anything in those cases)... let's just get basic type lookup working first.)

`ModuleLoader::ModuleImpl` will need to implement `reportResolution()` by appending to some list of all resolutions seen in the module.

The last question is, where do we actually put the identifier list in the schemas? `Node.SourceInfo` is associated with a specific node, but our identifier list will be for the whole file. I think there are two options:
1. We could say that the `Node.SourceInfo` corresponding to the file node gets the identifier list. All other nodes' `SourceInfo`s contain empty lists.
2. Define a new type `struct FileSourceInfo`, which contains file-level optional source information, and then place this in `CodeGeneratorRequest.RequestedFile`.

I think I slightly lean towards option 2 but I don't have a strong opinion.

Reply:
Thank you.  I also like option 2.
At first, I updated the `schema.capnp`. 
Does it align with your intentions?

Reply:
I added `reportResolution` for `ErrorReporter` interface and implement on `ModuleLoader::ModuleImpl`.
cf50a67c85ddd20230204c93cdf57c17f15ff5b8

Reply:
@kentonv 

I have a few comments:

1. **Resolutions Member in `ModuleLoader::ModuleImpl`**: I believe we need to introduce a `resolutions` member variable in `ModuleLoader::ModuleImpl` to store the list of resolutions.

2. **Getter for Resolutions**: We’ll also need a getter function for `ModuleLoader::ModuleImpl`. The signature could be something like:

   ```cpp
   kj::Vector<schema::CodeGeneratorRequest::RequestedFile::FileSourceInfo::Identifier::Reader> getResolutions();
   ```

   This method could be added as a virtual method to the `Module` class, allowing it to be accessed from the `CompilerMain` class.

3. **Schema's `FileSourceInfo`**: Should the schema’s `FileSourceInfo` be set in the `kj::MainBuilder::Validity generateOutput()` method in the `CompilerMain` class? Is this the correct approach?

Reply:
Hmm, I think the `resolutions` member of `ModuleImpl` should contain some C++ representation the resolutions, not Cap'n Proto representation, since allocating lots of little capnp messages would be inefficient. So we need a type like:

```
struct Resolution {
  uint32_t startByte;
  uint32_t endByte;

  struct Type { uint64_t typeId; };
  struct Member {
    uint64_t parentTypeId;
    uint16_t ordinal;
  };

  kj::OneOf<Type, Member>;
};
```

In fact this type should probably be the input type for `reportResolution()`, just to make things easier.

And then I think `getResolutions()` should return an array of that type, instead of trying to return struct readers:

```
kj::ArrayPtr<const Resolution> getResolutions();
```

Otherwise, everything you say makes sense, yes.

Reply:
@kentonv 
I modified the source code as you suggested.  
Currently, as a trial, I have implemented `errorReporter.reportResolution()` only in the `Expression::RELATIVE_NAME` case. Which other cases should I implement it for?

Reply:
Seems like it would be useful to report ABSOLUTE_NAME and IMPORT as well.

ABSOLUTE_NAME is when a name is prefixed by a `.` to indicate it should be looked up starting from the global scope.

IMPORT is, of course, `import` directives -- these will resolve to a file.

I think people will expect to be able to jump-to-definition on both of these.